source/disruption/line_reports_api.cpp

```diff
@@ -104,8 +101,56 @@ struct LineReport {
         pb_creator.fill(stop_areas, report->mutable_pt_objects(), 0);
         pb_creator.fill(stop_points, report->mutable_pt_objects(), 0, with_line_sections);
     }
+
+    nt::UrisList all_uris() const {
+        nt::UrisList uris;
+        uris.insert(line->uri);
+        for (auto& network : networks) {
+            uris.insert(network->uri);
+        }
+        for (auto& route : routes) {
+            uris.insert(route->uri);
+        }
+        for (auto& stop_area : stop_areas) {
+            uris.insert(stop_area->uri);
+        }
+        for (auto& stop_point : stop_points) {
+            uris.insert(stop_point->uri);
+        }
+        return uris;
+    }
 };
 
+void filter_excess_impacts_in_uri_filtering_mode(const std::string& filter,
+                                                 navitia::PbCreator& pb_creator,
+                                                 const std::vector<LineReport>& line_reports) {
+    if (!filter.empty()) {
+        nt::UrisList line_report_uris;
+        for (auto& line_report : line_reports) {
+            auto uris = line_report.all_uris();
+            line_report_uris.insert(uris.cbegin(), uris.cend());
+        }
+
+        auto erase_allowed = [](const nt::UrisList& uris, const nt::UrisList& impacted_uris) -> bool {
+            for (const auto& impacted_uri : impacted_uris) {
+                if (std::find(uris.cbegin(), uris.cend(), impacted_uri) == uris.cend()) {
+                    return true;
+                }
+            }
+            return false;
+        };
+
+        // Clean impacts if needed
+        for (auto it = pb_creator.impacts.begin(); it != pb_creator.impacts.end();) {
+            if (erase_allowed(line_report_uris, (*it).get()->informed_entities_uris())) {
+                it = pb_creator.impacts.erase(it);
+            } else {
+                ++it;
+            }
+        }
```

Comment:
If a protobuf repeated array is compliant with a `ForwardIterator`, you could just write:

```cpp
boost::remove_if(pb_creator.impacts, [&](const auto& impact){
    return erase_allowed(line_report_uris, impact.get()->informed_entities_uris());
});
```

Reply:
Good idea, but only the old way works with std::set... :cry: 
https://stackoverflow.com/questions/24263259/c-stdseterase-with-stdremove-if